text/0001-bee-network/0001-bee-network.md

```diff
@@ -0,0 +1,206 @@
+Each node can therefore have several neighboring nodes with which it communicates.
+In an Internet-of-Things environment, a node might need to understand different protocols  (TCP, UDP, Bluetooth, ...) in order to interact with its surroundings.
+This RFC proposes a generic, asynchronous networking interface which abstracts specific communication protocols. Furthermore it provides an easy and convenient way to exchange messages between neighboring nodes.
+
+# Motivation
+
+Each node can have several neighboring nodes. To send or receive messages from one node to another node, a networking interface is required.
+In the current state of the IOTA mainnet, nodes only accept TCP connections. This however could change, especially to suite the needs of Internet of Things (IoT).
+In an IoT environment for example, there could exist devices which listen to TCP, whereas others only have UDP or Bluetooth ports to connect.
+The goal of this RFC is to provide a generic networking interface. It should provide an relatively convenient way for nodes to peer and exchange information with its neighbors.
+Furthermore, this networking interface should be able to handle different types of connections at the same time, independent of their nature.
+
+# Detailed design
+
+It is assumed that all types of connection have certain similarities:
+
+- a connection needs to be opened / initialized
+- there must exist a functionality to send data over it
+- there must exist a functionality to read data from it
+- after a connection is no longer needed, it will get dissolved
+
+As a result, the following trait is defined:
+
+```rust
+pub trait Connection: Sized {
+
+    type InitConfig;
+    type Error;
+
+    fn connect(config: Self::InitConfig) -> Result<Self, Self::Error>;
+    fn send(&mut self, msg: Message) -> Result<(), Self::Error>;
+    fn recv(&mut self, size: u64) -> Result<Message, Self::Error>;
+    fn try_recv(&mut self) -> Option<Result<Message, Self::Error>>;
+    fn disconnect(self);
+
+}
+```
+
+As shown in the above `Connection` trait, each connection will be initialized by its `InitConfig`.
+This `InitConfig` contains all parameters that are needed to initialize/open up the connection.
+Furthermore, each connection contains its own `Error` type.
+
+**Example**: a TCP connection could be defined as follows:
+
+```rust
+pub struct TcpConnection {
+
+    stream: TcpStream,
+
+}
+
+impl Connection for TcpConnection {
+
+    type InitConfig = (IpAdress, Port);
+    type Error = ();
+
+    fn connect(config: Self::InitConfig) -> Result<Self, Self::Error> {
+                
+        let mut stream = TcpStream::connect(ip_address, port)?;
+        Ok(TcpConnection{ stream })
+        
+    }
+
+    fn send(&mut self, msg: Message) -> Result<(), Self::Error> {
+       self.stream.write(msg.bytes())?;
+    }
+
+    fn recv(&mut self, size: u64) -> Result<Message, Self::Error> {
+       self.stream.read(&mut [0; size])?;
```

Comment:
Arrays are stack objects and so must have a constant size. You'll need use a dynamically-sized type like `Vec` for this.